### PR TITLE
Protobuf3 serializer for Cluster and Cluster Tools

### PR DESCRIPTION
**Pull request by alexvaluyskiy**
Fixes https://github.com/akkadotnet/akka.net/issues/2485

### CODE REVIEW

**Comment by jeff-pang** [comment]
I want to try taking a stab at applying the Profobuf3 change to the coreclr branch. Is that ok?

File: src/contrib/cluster/Akka.Cluster.Tools/Client/Serialization/ClusterClientMessageSerializer.cs [hunk 1]
```diff
@@ -112,47 +109,23 @@ public override string Manifest(object o)
...
+            foreach (var contactPoint in message.ContactPoints)
             {
-                message.WriteTo(gzipStream);
-                gzipStream.Dispose();
-                return bos.ToArray();
+                protoMessage.ContactPoints.Add(contactPoint);
             }
-        }
-
-        private byte[] Decompress(byte[] bytes)
```
Comment (alexvaluyskiy) [inline src/contrib/cluster/Akka.Cluster.Tools/Client/Serialization/ClusterClientMessageSerializer.cs]: I'm not sure that we should have compress/decompress logic inside serializers. I think it should be done in the remote transport. But DotNetty does not support it at the moment

File: src/contrib/cluster/Akka.Cluster.Tools/PublishSubscribe/Internal/TopicMessages.cs [hunk 2]
```diff
@@ -441,6 +441,28 @@ public SendToOneSubscriber(object message)
         {
             Message = message;
         }
+
+        private bool Equals(SendToOneSubscriber other)
```
Comment (alexvaluyskiy) [inline src/contrib/cluster/Akka.Cluster.Tools/PublishSubscribe/Internal/TopicMessages.cs]: Added Equals to a remote message

File: src/core/Akka.API.Tests/CoreAPISpec.ApproveCluster.approved.txt [hunk 3]
```diff
@@ -330,4 +332,15 @@ namespace Akka.Cluster.Routing
         public int TotalInstances { get; }
         public string UseRole { get; }
     }
+}
+namespace Akka.Cluster.Serialization
+{
+    public class ClusterMessageSerializer : Akka.Serialization.Serializer
```
Comment (alexvaluyskiy) [inline src/core/Akka.API.Tests/CoreAPISpec.ApproveCluster.approved.txt]: Serializer should be public, but protobuf messages is not

File: src/core/Akka.Cluster/Serialization/ClusterMessageSerializer.cs [hunk 4]
```diff
@@ -0,0 +1,505 @@
...
+            {
+                return ClusterRouterPoolToByteArray(pool);
+            }
+
+            throw new ArgumentException($"Can't serialize object of type [{obj.GetType()}] in [{nameof(ClusterMessageSerializer)}]");
+        }
+
+        public override object FromBinary(byte[] bytes, Type type)
+        {
+            if (type == typeof(ClusterHeartbeatSender.Heartbeat))
```
Comment (alexvaluyskiy) [inline src/core/Akka.Cluster/Serialization/ClusterMessageSerializer.cs]: Todo: create types map here

File: src/core/Akka.Cluster/Serialization/ClusterMessageSerializer.cs [hunk 5]
```diff
@@ -0,0 +1,505 @@
...
+                (int)clusterRouterPoolSettingsProto.MaxInstancesPerNode,
+                clusterRouterPoolSettingsProto.AllowLocalRoutees,
+                clusterRouterPoolSettingsProto.UseRole == string.Empty ? null : clusterRouterPoolSettingsProto.UseRole);
+        }
+
+        //
+        // Gossip
+        //
+
+        private static Proto.Msg.Gossip GossipToProto(Gossip gossip)
```
Comment (alexvaluyskiy) [inline src/core/Akka.Cluster/Serialization/ClusterMessageSerializer.cs]: I should create a couple of nbench specs for complex messages

File: src/core/Akka.Cluster/Serialization/ClusterMessageSerializer.cs [hunk 6]
```diff
@@ -0,0 +1,505 @@
...
+        {
+            return new Address(
+                addressProto.Protocol,
+                addressProto.System,
+                addressProto.Hostname,
+                addressProto.Port == 0 ? null : (int?)addressProto.Port);
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private static Proto.Msg.UniqueAddress UniqueAddressToProto(UniqueAddress uniqueAddress)
```
Comment (alexvaluyskiy) [inline src/core/Akka.Cluster/Serialization/ClusterMessageSerializer.cs]: Is it make snce to use AggresiveInlining here?

File: src/core/Akka/Serialization/NewtonSoftJsonSerializer.cs [hunk 7]
```diff
@@ -217,7 +217,7 @@ protected override JsonProperty CreateProperty(MemberInfo member, MemberSerializ
         /// </summary>
         public override int Identifier
         {
-            get { return -3; }
+            get { return 50; }
```
Comment (alexvaluyskiy) [inline src/core/Akka/Serialization/NewtonSoftJsonSerializer.cs]: Negative numbers should not be used for serializer's identifiers

**Comment by Aaronontheweb** [comment]
@alexvaluyskiy quick question: is this going to cause any issues during upgrades for people who have some nodes on Protobuff2? I assume not, but just wanted to confirm.

**Pull request closed without merging**